gapis/trace/desktop/ggp_trace.go

```diff
@@ -0,0 +1,600 @@
+				children = append(children, p+":"+a)
+			}
+		}
+	}
+
+	dir, file := path.Split(uri)
+	name := file
+	if name == "" {
+		name = dir
+	}
+
+	tttn := &tracer.TraceTargetTreeNode{
+		Name:            name,
+		Icon:            nil,
+		URI:             uri,
+		TraceURI:        traceUri,
+		Children:        children,
+		Parent:          dir,
+		ApplicationName: "",
+		ExecutableName:  file,
+	}
+
+	return tttn, nil
+}
+
+func (t *GGPTracer) setPackageName(uri, name string) {
+	t.packageNameMutex.Lock()
+	defer t.packageNameMutex.Unlock()
+	t.packageNames[uri] = name
+}
+
+func (t *GGPTracer) getPackageName(uri string) (string, bool) {
+	t.packageNameMutex.RLock()
+	defer t.packageNameMutex.RUnlock()
+	if _, ok := t.packageNames[uri]; !ok {
+		return "", false
+	}
+	return t.packageNames[uri], true
+}
+
+// getListOutputColumns calls ggp list commands for the given listName, along
+// with extra arguments specified in extras. The contents of the given column
+// names will be returned. The content of each column will be represented in
+// a list of strings, in the same order shown in the ggp list output. And the
+// content of columns will be returned in the order of the column names.
+func getListOutputColumns(ctx context.Context, listName string, extras []string, columnNames ...string) ([][]string, error) {
+	ggpPath, err := ggp.GGPExecutablePath()
+	if err != nil {
+		return nil, log.Errf(ctx, err, "getting %v", listName)
+	}
+	executable := ggpPath.System()
+	args := []string{listName, "list"}
+	if len(extras) != 0 {
+		args = append(args, extras...)
+	}
+	cmd := shell.Command(executable, args...)
+	outBuf := &bytes.Buffer{}
+	errBuf := &bytes.Buffer{}
+	if err := cmd.Capture(outBuf, errBuf).Run(ctx); err != nil {
+		return nil, log.Errf(ctx, err, "run %v list getting command", listName)
+	}
+	t, err := ggp.ParseListOutput(outBuf)
+	if err != nil {
+		return nil, log.Errf(ctx, err, "parse %v list", listName)
+	}
+	result := make([][]string, len(columnNames))
+	for i, c := range columnNames {
+		l, err := t.ColumnByName(c)
+		if err != nil {
+			return nil, log.Errf(ctx, err, "getting %v %v(s)", listName, c)
+		}
+		result[i] = l
+	}
+	return result, nil
+}
+
+// parsePackageURI parses the project display name, package ID and application
+// string from the given URI in forms "package=<Package ID>:<Application>" or
+// "package=<Project Name>/<Package ID>:<Application>". The URI is allowed to
+// be partially complete, which means the URI may not be a complete URI
+// targeting to a valid tracing target, but must starts with "package=". If the
+// a field in the URI is missing, empty string will be returned for the
+// corresponding return value. One special case only used internally is:
+// "package=/", which returns proj="/", pkg="", app="".
+func parsePackageURI(ctx context.Context, uri string) (proj, pkg, app string, err error) {
+	re := regexp.MustCompile(`package=(\/$|[a-zA-Z0-9\_\-\s]+\/)?([a-z0-9\_\-\s]+)?(\:[a-zA-Z0-9\s\_\-]+$)?`)
```

Comment:
Nit: not sure regexp is the most maintainable approach here, but I guess it does the job.
Another approach would be to tokenize the URI using ['/', '=', ':'] as delimiters and analyse the resulting split.

Reply:
The allowable tokens are fairly well-defined, so I think the regex is okay at the moment. If we run into issues, we can split it out more carefully.